cpp/src/reader_c.cpp

```diff
@@ -0,0 +1,379 @@
+      *num_chunk_indices = output_indices.size();
+    } else {
+      *num_chunk_indices = 0;
+    }
+  } catch (...) {
+    *chunk_indices = nullptr;
+    *num_chunk_indices = 0;
+  }
+}
+
+void get_chunk(ChunkReaderHandle reader, int64_t chunk_index) {
+  if (!reader)
+    return;
+
+  try {
+    auto* cpp_reader = static_cast<ChunkReader*>(reader);
+    auto result = cpp_reader->get_chunk(chunk_index);
+    // Note: The C API doesn't return the result, so this is essentially a no-op
+    // In a real implementation, you might want to cache the result or pass it to a callback
+  } catch (...) {
+    // Silently handle errors as per the void API design
+  }
+}
+
+void get_chunks(ChunkReaderHandle reader,
+                const int64_t* chunk_indices,
+                size_t num_indices,
+                int64_t parallelism,
+                ArrowArray** arrays,
+                size_t* num_arrays) {
+  if (!reader || !chunk_indices || num_indices == 0 || !arrays || !num_arrays) {
+    if (arrays)
+      *arrays = nullptr;
+    if (num_arrays)
+      *num_arrays = 0;
+    return;
+  }
+
+  try {
+    auto* cpp_reader = static_cast<ChunkReader*>(reader);
+    std::vector<int64_t> indices(chunk_indices, chunk_indices + num_indices);
+
+    auto result = cpp_reader->get_chunks(indices, parallelism);
+    if (!result.ok()) {
+      *arrays = nullptr;
+      *num_arrays = 0;
+      return;
+    }
+
+    const auto& record_batches = result.ValueOrDie();
+    if (record_batches.empty()) {
+      *arrays = nullptr;
+      *num_arrays = 0;
+      return;
+    }
+
+    // Convert RecordBatches to Arrow C ABI arrays
+    *arrays = static_cast<ArrowArray*>(malloc(sizeof(ArrowArray) * record_batches.size()));
+    if (*arrays) {
+      *num_arrays = record_batches.size();
+      // TODO: Implement RecordBatch to ArrowArray conversion
+      // This requires converting Arrow C++ RecordBatch to Arrow C ABI format
+      // For now, just initialize the arrays to avoid crashes
+      for (size_t i = 0; i < record_batches.size(); ++i) {
+        memset(&(*arrays)[i], 0, sizeof(ArrowArray));
+      }
+    } else {
+      *num_arrays = 0;
+    }
+  } catch (...) {
+    *arrays = nullptr;
+    *num_arrays = 0;
+  }
+}
+
+void chunk_reader_destroy(ChunkReaderHandle reader) {
+  if (reader) {
+    delete static_cast<ChunkReader*>(reader);
+  }
+}
+
+// ==================== Reader C Implementation ====================
+
+ReaderHandle reader_new(FileSystemHandle fs,
+                        char* manifest,
+                        ArrowSchema* schema,
+                        const char* const* needed_columns,
+                        size_t num_columns,
+                        const ::ReadProperties* properties) {
+  // TODO: Implement reader creation
+  // This function has a void return type but needs to somehow return the created reader
+  // The API design seems incomplete - missing a ReaderHandle* parameter
+  if (!fs || !manifest || !schema || !properties) {
```

Comment:
`assert` replace the `return nullptr`? i don't think caller will pass a nullptr in those field